executor/include/executor/best_fit.hpp

```diff
@@ -0,0 +1,118 @@
+  static bool availability_check(const Executor<Properties...>& exec) {
+    return executor::is_executor_available_v<Executor>;
+  }
+
+  template <typename Executor, typename executor::instance_of_base<
+                                   Executor, executor::inline_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::instance_of_base<
+                                   Executor, executor::sse_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::instance_of_base<
+                                   Executor, executor::omp_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+
+  template <typename Executor, typename executor::instance_of_base<
+                                   Executor, executor::cuda_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
+};
+
+template <typename RuntimeChecks = executor_runtime_checks, typename Function,
+          typename... SupportedExecutors>
+void enable_exec_with_priority(
+    Function f, std::tuple<SupportedExecutors...> supported_execs) {
+  static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
+                "Runtime checks should inherit from executor_runtime_checks");
+  bool executor_selected = false;
+  executor::for_each_tuple_until(supported_execs, [&](auto& exec) {
+    if (RuntimeChecks::availability_check(exec) && RuntimeChecks::check(exec)) {
+      f(exec);
+      return (executor_selected = true);
+    }
+    return false;
+  });
+  // TODO: Throw warning or assert
+}
+
+template <typename RuntimeChecks = executor_runtime_checks, typename Function,
+          typename... SupportedExecutors>
+void enable_exec_with_priority(Function f, SupportedExecutors... execs) {
+  enable_exec_with_priority<RuntimeChecks>(f, std::make_tuple(execs...));
+}
+
+template <typename RuntimeChecks = executor_runtime_checks, typename Function,
+          typename... SupportedExecutors>
+void enable_exec_on_desc_priority(
+    Function f, std::tuple<SupportedExecutors...> supported_execs) {
+  static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
+                "Runtime checks should inherit from executor_runtime_checks");
+  auto num_supported_execs = std::tuple_size<decltype(supported_execs)>::value;
+  auto exec_is_supported = [&](auto& best_fit_exec) {
+    if (!RuntimeChecks::availability_check(best_fit_exec)) return false;
+    auto count = 0;
+    for_each_tuple_until(supported_execs, [&](auto& supported_exec) {
+      if (executor::is_same_template<decltype(best_fit_exec),
+                                     decltype(supported_exec)>::value &&
+          RuntimeChecks::check(supported_exec)) {
+        f(supported_exec);
+        return true;
+      }
+      ++count;
+      return false;
+    });
```

Comment:
What the difference between `enable_exec_with_priority` and `enable_exec_on_desc_priority`. It  seems to be a matter of order on how executors are picked? The former follows the natural order from `supported_execs` that latter follows the order of `best_fit_executors`. 

We might be able to reduce some of the code duplication between both methods.

Reply:
So the idea behind this is that both the functions `enable_exec_with_priority` and `enable_exec_on_desc_priority` choose from one of the executors passed into the function.

The difference lies that `enable_exec_with_priority` checks whether the executors should be chosen in the order the user passed the executors.

`enable_exec_on_desc_priority` checks which executor to be chosen in the order specified by the library i.e. omp > see > inline.

I tried to combine the common functionality but was not able to do so, since we can't modify tuples so re-ordering is not possible of the passed in executors.

Reply:
>  since we can't modify tuples so re-ordering is not possible of the passed in executors.

This is how I'm seeing it happen (pseudo-code)

```cpp
auto enable_exec_with_priority(sequence_of_executors_to_try)
{
   // tests the order specified in the sequence
  [...]
}

auto enable_exec_on_desc_priority()
{
    //create tuple with default order established by the library
    sequence_of_executors_to_try = std::tuple<...,...,...>;
    enable_exec_with_priority(sequence_of_executors_to_try);
}
```

is something like this not possible?

Reply:
One point missing was, that `enable_exec_on_desc_priority` also requires to pass the executors to try as the sequence of executors passed tells us which executors are supported by the function we are trying to call i.e. which executor types have overload available for them.
```cpp
auto enable_exec_on_desc_priority(sequence_of_executors_to_try)
```

So, what we want is this:
```cpp
auto enable_exec_on_desc_priority(sequence_of_executors_to_try)
{
    //create tuple by rearranging sequence_of_executors_to_try into the order specified by the library
    new_sequence_of_executors_to_try = std::tuple<...,...,...>;
    enable_exec_with_priority(new_sequence_of_executors_to_try);
}
```

I have not figured out a way to do this till now.

Reply:
> One point missing was, that `enable_exec_on_desc_priority` also requires to pass the executors to try as the sequence of executors passed tells us which executors are supported by the function we are trying to call i.e. which executor types have overload available for them.

Why is this required? Do we not have a mechanism which allows querying if a given method has a specific executor implementation?

Reply:
>  Do we not have a mechanism which allows querying if a given method has a specific executor implementation?
No, I couldn't get this to work.

Reply:
We should chat about this at some point then. It's an important feature to have but likely not before GSoC is done.

Reply:
Yeah, tracking it in #23